Example code in README.md raises errors in mypy

### Things to check first

- [X] I have searched the existing issues and didn't find my bug already reported there

- [X] I have checked that my bug is still present in the latest release


### Exceptiongroup version

1.2.0

### Python version

3.11.5

### What happened?

```
$ cat foo.py
from exceptiongroup import ExceptionGroup, catch

def value_key_err_handler(excgroup: ExceptionGroup) -> None:
    for exc in excgroup.exceptions:
        print('Caught exception:', type(exc))

def runtime_err_handler(exc: ExceptionGroup) -> None:
    print('Caught runtime error')

with catch({
    (ValueError, KeyError): value_key_err_handler,
    RuntimeError: runtime_err_handler
}):
    ...


$ mypy foo.py
foo.py:11: error: Dict entry 0 has incompatible type "tuple[type[ValueError], type[KeyError]]": "Callable[[ExceptionGroup[Any]], None]"; expected "type[BaseException] | Iterable[type[BaseException]]": "Callable[[BaseException], Any]"  [dict-item]
foo.py:12: error: Dict entry 1 has incompatible type "type[RuntimeError]": "Callable[[ExceptionGroup[Any]], None]"; expected "type[BaseException] | Iterable[type[BaseException]]": "Callable[[BaseException], Any]"  [dict-item]
Found 2 errors in 1 file (checked 1 source file)
```

I expected no errors

### How can we reproduce the bug?

mypy version 1.7.0, no configuration.

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -3,6 +3,11 @@ Version history
 
 This library adheres to `Semantic Versioning 2.0 <http://semver.org/>`_.
 
+**UNRELEASED**
+
+- Corrected the type annotation of the exception handler callback to accept a
+  ``BaseExceptionGroup`` instead of ``BaseException``
+
 **1.2.0**
 
 - Added special monkeypatching if `Apport <https://github.com/canonical/apport>`_ has

--- a/README.rst
+++ b/README.rst
@@ -54,7 +54,7 @@ containing more matching exceptions.
 
 Thus, the following Python 3.11+ code:
 
-.. code-block:: python3
+.. code-block:: python
 
     try:
         ...
@@ -66,7 +66,7 @@ Thus, the following Python 3.11+ code:
 
 would be written with this backport like this:
 
-.. code-block:: python3
+.. code-block:: python
 
     from exceptiongroup import BaseExceptionGroup, catch
 
@@ -91,7 +91,9 @@ Suppressing exceptions
 
 This library contains a backport of the ``contextlib.suppress()`` context manager from
 Python 3.12.1. It allows you to selectively ignore certain exceptions, even when they're
-inside exception groups::
+inside exception groups:
+
+.. code-block:: python
 
     from exceptiongroup import suppress
 

--- a/README.rst
+++ b/README.rst
@@ -68,13 +68,13 @@ would be written with this backport like this:
 
 .. code-block:: python3
 
-    from exceptiongroup import ExceptionGroup, catch
+    from exceptiongroup import BaseExceptionGroup, catch
 
-    def value_key_err_handler(excgroup: ExceptionGroup) -> None:
+    def value_key_err_handler(excgroup: BaseExceptionGroup) -> None:
         for exc in excgroup.exceptions:
             print('Caught exception:', type(exc))
 
-    def runtime_err_handler(exc: ExceptionGroup) -> None:
+    def runtime_err_handler(exc: BaseExceptionGroup) -> None:
         print('Caught runtime error')
 
     with catch({

--- a/src/exceptiongroup/_catch.py
+++ b/src/exceptiongroup/_catch.py
@@ -11,7 +11,7 @@ if sys.version_info < (3, 11):
     from ._exceptions import BaseExceptionGroup
 
 if TYPE_CHECKING:
-    _Handler = Callable[[BaseExceptionGroup], Any]
+    _Handler = Callable[[BaseExceptionGroup[Any]], Any]
 
 
 class _Catcher:

--- a/src/exceptiongroup/_catch.py
+++ b/src/exceptiongroup/_catch.py
@@ -11,7 +11,7 @@ if sys.version_info < (3, 11):
     from ._exceptions import BaseExceptionGroup
 
 if TYPE_CHECKING:
-    _Handler = Callable[[BaseException], Any]
+    _Handler = Callable[[BaseExceptionGroup], Any]
 
 
 class _Catcher:

